[SPARK-7249] Updated Hadoop dependencies due to inconsistency in the versions

FavioVazquez opened the pull request:
Updated Hadoop dependencies due to inconsistency in the versions. Now the global properties are the ones used by the hadoop-2.2 profile, and the profile was set to empty but kept for backwards compatibility reasons.

Changes proposed by @vanzin resulting from previous pull-request https://github.com/apache/spark/pull/5783 that did not fixed the problem correctly.

Please let me know if this is the correct way of doing this, the comments of @vanzin are in the pull-request mentioned.

FavioVazquez committed:
- Updated protobuf-java version of com.google.protobuf dependancy to fix blocking error when connecting to HDFS via the Hadoop Cloudera HDFS CDH5 (fix for 2.5.0-cdh5.3.3 version)

FavioVazquez committed:
- Updated Hadoop dependencies due to inconsistency in the versions. Now the global properties are the ones used by the hadoop-2.2 profile, and the profile was set to empty but kept for backwards compatibility reasons

FavioVazquez committed:
- Updated Hadoop dependencies due to inconsistency in the versions. Now the global properties are the ones used by the hadoop-2.2 profile, and the profile was set to empty but kept for backwards compatibility reasons

AmplabJenkins commented:
Can one of the admins verify this patch?

vanzin commented on pom.xml:
```diff
@@ -1629,17 +1629,21 @@
-        <hbase.version>0.98.7-hadoop2</hbase.version>
-        <avro.mapred.classifier>hadoop2</avro.mapred.classifier>
-        <codehaus.jackson.version>1.9.13</codehaus.jackson.version>
+        <hadoop.version>1.2.0</hadoop.version>
```
I don't think `1.2.0` is a valid version. The previous value used to be `1.0.4`.

srowen replied:
None of this should actually affect the releases people use since they build with profiles explicitly enabled as desired. It's a good change of defaults, for hygiene though.

The story here is that Spark 1.2 was sorta accidentally released as depending (in Maven) on 2.2, and so that was expressed later, but this is the logical conclusion of that process. The default build which you'd get if approaching this as a developer is and should be a consistent Hadoop 2.2 build.

However more needs to change besides this. The release script has to set a "hadoop-1" profile in some cases now. The docs about building vs various Hadoop versions has to be updated too.

vanzin commented:
This LGTM, just a small issue left behind. There's some cleanup that could be done in the other `hadoop-2.x` profiles now, since they contain mostly redundant stuff.

@srowen could you trigger tests for this? Thanks!

tgravescs commented:
jenkins, test this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31337 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31337/consoleFull) for   PR 5786 at commit [`7e9955d`](https://github.com/apache/spark/commit/7e9955df29b5d5c9cda950636d51da753e6d17ea).

srowen commented:
(PS update the title please per https://cwiki.apache.org/confluence/display/SPARK/Contributing+to+Spark )

SparkQA commented:
[Test build #31337 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31337/consoleFull) for   PR 5786 at commit [`7e9955d`](https://github.com/apache/spark/commit/7e9955df29b5d5c9cda950636d51da753e6d17ea).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch **adds the following new dependencies:**
   * `jackson-core-asl-1.9.13.jar`
   * `jackson-jaxrs-1.9.13.jar`
   * `jackson-mapper-asl-1.9.13.jar`
   * `jackson-xc-1.9.13.jar`
   * `protobuf-java-2.5.0.jar`

 * This patch **removes the following dependencies:**
   * `jackson-core-asl-1.8.8.jar`
   * `jackson-jaxrs-1.8.8.jar`
   * `jackson-mapper-asl-1.8.8.jar`
   * `jackson-xc-1.8.8.jar`
   * `protobuf-java-2.4.1.jar`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31337/
Test FAILed.

FavioVazquez commented:
Can someone please explain me what failed? I'm kinda new to this, and I'm not sure what failed, I want to know to fix it and maybe create another pull request. Thanks

vanzin commented:
You can click on the logs to see the failure; it's a pyspark unit tests, I'd be surprised if your change caused it.

FavioVazquez commented:
In the link https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31337/, there is no failure in the tests, but in the console output the error thrown was this:

FAIL: test_count_by_value_and_window (__main__.WindowFunctionTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "pyspark/streaming/tests.py", line 418, in test_count_by_value_and_window
    self._test_func(input, func, expected)
  File "pyspark/streaming/tests.py", line 133, in _test_func
    self.assertEqual(expected, result)
AssertionError: Lists differ: [[1], [2], [3], [4], [5], [6], [6], [6], [6], [6]] != [[1], [2], [3], [4], [5], [6], [6], [6], [6]]

First list contains 1 additional elements.
First extra element 9:
[6]

- [[1], [2], [3], [4], [5], [6], [6], [6], [6], [6]]
?                                          -----

+ [[1], [2], [3], [4], [5], [6], [6], [6], [6]]

----------------------------------------------------------------------
Ran 40 tests in 134.429s

FAILED (failures=1)
('timeout after', 20)
('timeout after', 20)
('timeout after', 20)
('timeout after', 5)
Had test failures; see logs.
[error] Got a return code of 255 on line 240 of the run-tests script.
Archiving unit tests logs...

So I'm not sure what happened.

srowen commented:
Yes I imagine this is an unrelated failure, so let's try again. however I'm also sure that this change will require changes in a number of other places, like the docs, the release script, and the Jenkins config. We'll have to coordinate that a bit. Looping in a busy @pwendell -- better to get this in early for 1.4.0? feels like we should.

srowen commented:
Jenkins, retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31400 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31400/consoleFull) for   PR 5786 at commit [`7e9955d`](https://github.com/apache/spark/commit/7e9955df29b5d5c9cda950636d51da753e6d17ea).

SparkQA commented:
[Test build #31400 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31400/consoleFull) for   PR 5786 at commit [`7e9955d`](https://github.com/apache/spark/commit/7e9955df29b5d5c9cda950636d51da753e6d17ea).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch **adds the following new dependencies:**
   * `jackson-core-asl-1.9.13.jar`
   * `jackson-jaxrs-1.9.13.jar`
   * `jackson-mapper-asl-1.9.13.jar`
   * `jackson-xc-1.9.13.jar`
   * `jaxb-api-2.2.7.jar`
   * `jaxb-core-2.2.7.jar`
   * `jaxb-impl-2.2.7.jar`
   * `pmml-agent-1.1.15.jar`
   * `pmml-model-1.1.15.jar`
   * `pmml-schema-1.1.15.jar`
   * `protobuf-java-2.5.0.jar`

 * This patch **removes the following dependencies:**
   * `activation-1.1.jar`
   * `jackson-core-asl-1.8.8.jar`
   * `jackson-jaxrs-1.8.8.jar`
   * `jackson-mapper-asl-1.8.8.jar`
   * `jackson-xc-1.8.8.jar`
   * `jaxb-api-2.2.2.jar`
   * `jaxb-impl-2.2.3-1.jar`
   * `protobuf-java-2.4.1.jar`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31400/
Test PASSed.

FavioVazquez commented:
Hi, now that the test have passed what will happen now? I've read the documentation in https://cwiki.apache.org/confluence/display/SPARK/Contributing+to+Spark#ContributingtoSpark-PullRequest but I'm not sure, will this be merge to master? Should I do something else? keep the PR open?

Thanks.

srowen commented:
You can see my comments here? there is more to do on this change I think before it can be merged.

FavioVazquez commented:
Oh I see. Thanks. Should I keep the PR open then?

srowen commented:
Yes, you just push more changes to the branch.

FavioVazquez committed:
- Cleanup in hadoop-2.x profiles since they contained mostly redundant stuff.

FavioVazquez committed:
- Fixed hadoop-1 version to match jenkins build profile in hadoop1.0 tests and documentation

FavioVazquez committed:
- Updated documentation about specifying the hadoop version in building-spark. Now is clear that Spark will build against Hadoop 2.2.0 by default.
- Added Cloudera CDH 5.3.3 without MapReduce example in the building-spark doc.

FavioVazquez committed:
- Fixed typo in the make-distribution.sh file and added hadoop-1 in the Related profiles

FavioVazquez commented:
Are this changes what is needed to complete the patch?

vanzin commented on docs/building-spark.md:
```diff
@@ -83,6 +83,10 @@ mvn -Dhadoop.version=1.2.1 -DskipTests clean package
 # Cloudera CDH 4.2.0 with MapReduce v1
 mvn -Dhadoop.version=2.0.0-mr1-cdh4.2.0 -DskipTests clean package
 
+#Cloudera CDH 5.3.3 without MapReduce
```
I would not bother adding this section.

vanzin commented on pom.xml:
```diff
@@ -1656,12 +1660,12 @@
       <properties>
         <hadoop.version>2.4.0</hadoop.version>
-        <protobuf.version>2.5.0</protobuf.version>
+	<protobuf.version>${protobuf.version}</protobuf.version>
```
Just delete this line instead. No point in setting the variable to the value it already has. Same in other places.

srowen replied:
Yeah, these other profiles can become simpler now.

srowen commented on pom.xml:
```diff
@@ -1629,26 +1629,30 @@
     </profile>
 
     <profile>
+      <id>hadoop-2.2</id>
```
I'd comment why this is empty.

srowen commented on make-distribution.sh:
```diff
@@ -57,7 +57,7 @@ while (( "$#" )); do
       echo "Error: '--hadoop' is no longer supported:"
       echo "Error: use Maven profiles and options -Dhadoop.version and -Dyarn.version instead."
-      echo "Error: Related profiles include hadoop-0.23, hdaoop-2.2, hadoop-2.3 and hadoop-2.4."
+      echo "Error: Related profiles include hadoop-0.23, hadoop-1, hadoop-2.2, hadoop-2.3 and hadoop-2.4."
```
Good catch on the spelling; I think we don't need to mention `hadoop-2.2` now as its a no-op.

srowen commented:
The `create-release.sh` script will need to change now too in this PR. Keeping @pwendell in the loop.

FavioVazquez commented:
Thanks for your comments @vanzin @srowen and changing the files with your suggestions and push the changes soon

FavioVazquez committed:
- Simplified Hadoop profiles due to new setting of global properties in the pom.xml file
- Added comment to specify that the hadoop-2.2 profile is now the default hadoop profile in the pom.xml file
- Erased hadoop-2.2 from related hadoop profiles now that is a no-op in the make-distribution.sh file

srowen commented:
Getting there but there's still more to change to fully button this up.

- As Marcelo notes, we shouldn't add a CDH5-specific note; it's not needed
- `create-release.sh` needs to change around line 230 to update how the releases are made
- In fact, `create-release.sh` has more instances of `-Phadoop-2.2 -Dhadoop.version=2.2.0` that can go
- Same for `run-tests` and `scalastyle` in `dev/`
- `hadoop-third-party-distributions.md` could use a better example than Hadoop 2.2 then
- `building-spark.md` has another instance where `-Phadoop-2.2 -Dhadoop.version=2.2.0` can be removed

Then we'll need to change Jenkins config promptly. @shaneknapp what do you think about that? if this merges, then some of the Hadoop 1 builds need to be updated. Just a change to their flags.

FavioVazquez commented:
Great, I'll do that right away

vanzin commented on pom.xml:
```diff
@@ -1629,26 +1629,28 @@
+        <hadoop.version>1.0.4</hadoop.version>
+        <protobuf.version>2.4.1</protobuf.version>
+        <hbase.version>0.98.7-hadoop1</hbase.version>
+        <avro.mapred.classifier>hadoop1</avro.mapred.classifier>
```
hmm... the default value is empty, which tells me this should be empty.

Also, on the same note, I wonder if the default now should be `hadoop2`.

vanzin replied:
This needs to be empty, not `hadoop1`.

srowen replied:
Yeah @vanzin I couldn't recall why that was left empty, but according to https://issues.apache.org/jira/browse/SPARK-4359 it actually shouldn't be empty. So I think this is actually a good change.

vanzin replied:
Thing is, there is not "hadoop1" jar in http://repo1.maven.org/maven2/org/apache/avro/avro-mapred/1.7.6/.

srowen replied:
I see it there? http://repo1.maven.org/maven2/org/apache/avro/avro-mapred/1.7.6/avro-mapred-1.7.6-hadoop1.jar

I looked at the POM and that's how it sets the classifier too. Looks like the no-classifier version is the same as `hadoop1` anyway.

vanzin replied:
Man, I swear I missed it when I looked at that page. Anyway, since it exists, `hadoop1` is fine.

vanzin commented:
Since you're making all the extra changes, let me suggest another one. In `yarn/pom.xml`, there's a `hadoop-2.2` profile that should be enabled when the Hadoop version is `2.2.0`, which is now the default. I'd add this code there, in the profile:

      <activation>
        <property>
          <name>hadoop.version</name>
          <value>2.2.0<value>
        </property>
      </activation>

Those building against 2.2.x or some distro-specific variation of that string would still have to manually enable `-Phadoop-2.2`, though.

shaneknapp commented:
what exactly will need to be changed in the build configs?  there's a lot of places that hadoop is specified:  multi-project axis configs, maven opts (for other projects that use spark), etc.

depending on the extent of the config wrangling, i might want to have a brief downtime to ensure i do it right.  :)

FavioVazquez committed:
- Erased unnecessary CDH5-specific note in docs/building-spark.md
- Remove example of instance -Phadoop-2.2 -Dhadoop.version=2.2.0 in docs/building-spark.md
- Enabled hadoop-2.2 profile when the Hadoop version is 2.2.0, which is now the default .Added comment in the yarn/pom.xml to specify that.

FavioVazquez committed:
- Default value of avro.mapred.classifier has been set to hadoop2 in pom.xml
- Cleaned up hadoop-2.3 and 2.4 profiles due to change in the default set in avro.mapred.classifier in pom.xml

FavioVazquez committed:
- Erased unnecessary instance of -Phadoop-2.2 -Dhadoop.version=2.2.0 in create-release.sh
- Updated how the releases are made in the create-release.sh no that the default hadoop version is the 2.2.0
- Erased unnecessary instance of -Phadoop-2.2 -Dhadoop.version=2.2.0 in scalastyle
- Erased unnecessary instance of -Phadoop-2.2 -Dhadoop.version=2.2.0 in run-tests
- Better example given in the hadoop-third-party-distributions.md now that the default hadoop version is 2.2.0

FavioVazquez commented:
Please check if this are the changes that you suggested @vanzin @srowen.

Thanks.

srowen commented on dev/create-release/create-release.sh:
```diff
@@ -228,8 +228,8 @@ if [[ ! "$@" =~ --skip-package ]]; then
   # share the same Zinc server.
-  make_binary_release "hadoop1" "-Phive -Phive-thriftserver -Dhadoop.version=1.0.4" "3030" &
-  make_binary_release "hadoop1-scala2.11" "-Phive -Dscala-2.11" "3031" &
+  make_binary_release "hadoop2" "-Phive -Phive-thriftserver -Dhadoop.version=2.2.0" "3030" &
```
The names of the releases won't change, just their 'implementation'. This would have to select the Hadoop 1 profile and doesn't need to set version. Likewise for the next one.

srowen commented on dev/create-release/create-release.sh:
```diff
@@ -118,14 +118,14 @@ if [[ ! "$@" =~ --skip-publish ]]; then
 
-  build/mvn -DskipTests -Dhadoop.version=2.2.0 -Dyarn.version=2.2.0 \
-    -Pyarn -Phive -Phive-thriftserver -Phadoop-2.2 -Pspark-ganglia-lgpl -Pkinesis-asl \
+  build/mvn -DskipTests -Dyarn.version=2.2.0 \
```
Ah `yarn.version` isn't needed either as it defaults to `hadoop.version`

vanzin commented on yarn/pom.xml:
```diff
@@ -109,6 +109,16 @@
       </properties>
+      <!-- Enable hadoop-2.2 profile when the Hadoop version is 2.2.0, which is now 
+      the default. Those building against 2.2.x or some distro-specific variation of 
+      that string would still have to manually enable -Phadoop-2.2.
```
nit: drop "would".

vanzin commented:
Still one small thing left on the pom side, rest looks good. Will let other comment about the release scripts since I'm not familiar with those.

srowen commented on yarn/pom.xml:
```diff
@@ -109,6 +109,16 @@
+      -->
+      <activation>
+       <property>
+        <name>hadoop.version</name>
```
This gets kinda circular since the profile also sets this version. I think we've had profiles set versions instead of the other way around. This profile is in theory not just for "2.2.0"

vanzin replied:
No, this profile is specific to 2.2.0. Other 2.x versions don't need it.

srowen replied:
What about a theoretical "2.2.1"? In other similar profiles we haven't set it up this way so I wouldn't add this purely for consistency.

vanzin replied:
That's what the comment addresses.

The issue here is that if you just run "mvn -Pyarn test", the YARN tests will fail. So we want this profile to be activated to avoid that.

I guess one solution is to use `<activeByDefault>true</activeByDefault>`. My peeve with that approach is that the name is incredibly misleading - the above command line would fail the same way, because this property just activates the profile if no other profile has been manually activated.

srowen replied:
Can we just fold this config into the main POM config for yarn? 2.2 is the default after all, and I don't think YARN support works at all with Hadoop 1 at this point anyway. I could be missing something.

vanzin replied:
> Can we just fold this config into the main POM config for yarn?

I guess that could work. If there's any discrepancy in the versions in other hadoop profiles, we can override the versions later, but adding the dependencies for non-hadoop-2.2 is at worst redundant.

FavioVazquez commented:
Yes I saw the hadoop1.jar there also. Should I keep it like this or change the hadoop-1 profile to the hadoop2 also?

FavioVazquez commented:
Great. My confusion is with the create-release, what have to be change in there? So the names of the releases should be kept but change the implementation? I'm not quite sure what you ment there @srowen

srowen commented:
The `hadoop-1` profile needs the `hadoop1` classifier.
In `create-release.sh` the first two builds need to be, I think:

```
  make_binary_release "hadoop1" "-Phive -Phive-thriftserver -Phadoop-1" "3030" &
  make_binary_release "hadoop1-scala2.11" "-Phive -Phadoop-1 -Dscala-2.11" "3031" &
```

FavioVazquez commented:
Yes I figured it out a second before you told me looking at the other implementations. Thanks. I'll be pushing the changes in a second

FavioVazquez committed:
- Updated hadoop1 releases in create-release.sh  due to changes in the default hadoop version set
- Erased unnecessary instance of -Dyarn.version=2.2.0 in create-release.sh
- Prettify comment in yarn/pom.xml

FavioVazquez commented:
Please let me now if this completes the changes in the patch. Thank you.

srowen commented:
Looking strong; I'm still wondering about how to handle `yarn`. give @vanzin a while to respond but my idea was just to merge this entire profile's config into the main profile if indeed 2.2 is the default, and for YARN, nothing earlier is supported anyway.

FavioVazquez commented:
Great @srowen. Lets wait for @vanzin comments on this

vanzin commented:
I already did but github will make it really hard for you to find it. :-/ Here's the comment:

============
>  Can we just fold this config into the main POM config for yarn?

I guess that could work. If there's any discrepancy in the versions in other hadoop profiles, we can override the versions later, but adding the dependencies for non-hadoop-2.2 is at worst redundant.

FavioVazquez commented:
OK @vanzin so I should change the yarn confing in the main POM and leave the yarn/pom.xml as it was?

vanzin commented:
No; basically the `<dependency>` section of the `hadoop-2.2` profile in the YARN pom should be moved to the `yarn` profile in the root pom; and the profile in the YARN pom can be deleted.

FavioVazquez commented:
I see. I'll do that right away

FavioVazquez committed:
- Moved <dependencies> <activation> and <properties> sections of the hadoop-2.2 profile in the YARN POM to the YARN profile in the root POM
- Erased unnecessary hadoop-2.2 profile from the YARN POM

FavioVazquez commented:
@vanzin I think that's what you ment. Check if everithing is OK now please.

vanzin commented on pom.xml:
```diff
@@ -1671,6 +1668,57 @@
+      <properties>
+        <jersey.version>1.9</jersey.version>
+      </properties>
+      <!-- Enable hadoop-2.2 profile when the Hadoop version is 2.2.0, which is now 
```
This comment (and the `<activation>` part) should not be added here. The `yarn` profile is never automatically enabled, and we need to keep it that way.

vanzin commented:
Almost there, just need to delete some stuff that shouldn't be added to the root pom. I'll defer to @srowen on the other parts.

FavioVazquez committed:
- Deleted unnecessary comments and <activation> tag on the YARN profile in the main POM

FavioVazquez commented:
You are right @vanzin, I just pushed those changes. I'll wait for @srowen comments

srowen commented:
The `yarn` profile just exists to turn on these modules, right? I meant that the stuff in the `yarn` profile in `yarn/pom.xml` should be pulled into the main config of `yarn/pom.xml`, not `pom.xml`. Isn't that easier (keeps YARN stuff in the `yarn` module) or do I miss something?

This is turning into a good cleanup.

vanzin commented:
>  I meant that the stuff in the yarn profile in yarn/pom.xml should be pulled into the main config of yarn/pom.xml

D'oh. Yeah, that sounds like a better idea.

FavioVazquez commented:
So should I erase the yarn profile from the root POM and move the entire profile into the yarn/POM?

srowen commented:
No, the parent should contain the same `yarn` profile which only activates the YARN-related modules. The content of the `hadoop-2.2` profile that was in `yarn/pom.xml` should simply be integrated into `yarn/pom.xml`. Then that `hadoop-2.2` profile can go away.

FavioVazquez committed:
- Cleaned up the main POM concerning the yarn profile
- Erased hadoop-2.2 profile from yarn/pom.xml and its content was integrated into yarn/pom.xml

FavioVazquez commented:
Hope that's what you have been talking about @srowen @vanzin

vanzin commented on yarn/pom.xml:
```diff
@@ -97,59 +98,44 @@
   </dependencies>
-
-  <!--
-    See SPARK-3710. hadoop-yarn-server-tests in Hadoop 2.2 fails to pull some needed
```
Yes, this is the right idea, but could you keep the comment mentioning the bug for which these dependencies were added?

vanzin commented:
Just a comment about the comment you deleted, which I think should be maintained. Thanks for the patience with us in this patch!

FavioVazquez committed:
- Added comment related to SPARK-3710 about  hadoop-yarn-server-tests in Hadoop 2.2 that fails to pull some needed dependencies

FavioVazquez commented:
@vanzin @srowen thank you for helping me a lot with this patch. Let me know if this finishes the patch please.

srowen commented:
I noticed a few more things that need to be updated. Kind of confusingly, the Spark build conflates Hadoop 1.0 with Hadoop 2.0 and 2.1, so you'll need to add `-Phadoop-1` to a few more places. All of these need it, I think:

create-release.sh
  make_binary_release "cdh4" "-Phive -Phive-thriftserver -Dhadoop.version=2.0.0-mr1-cdh4.2.0" "3032" &

building-spark.md

# Apache Hadoop 1.2.1
mvn -Dhadoop.version=1.2.1 -DskipTests clean package

# Cloudera CDH 4.2.0 with MapReduce v1
mvn -Dhadoop.version=2.0.0-mr1-cdh4.2.0 -DskipTests clean package

dev/run-tests

if [ "$AMPLAB_JENKINS_BUILD_PROFILE" = "hadoop1.0" ]; then
  export SBT_MAVEN_PROFILES_ARGS="-Dhadoop.version=1.0.4"
elif [ "$AMPLAB_JENKINS_BUILD_PROFILE" = "hadoop2.0" ]; then
  export SBT_MAVEN_PROFILES_ARGS="-Dhadoop.version=2.0.0-mr1-cdh4.1.1"


Soon I want to get rid of this unsupported CDH4 info/profile (and CDH*3* docs! surely Spark hasn't worked with that in a long time). Separate issue.

srowen commented:
@shaneknapp if this goes in, then the essence of the config change is this: anything building for Hadoop 1.x, 2.0, or 2.1 now needs `-Phadoop-1` to maintain the same config. Anything that used to specify `-Phadoop-2.2 -Dhadoop.version=2.2.0` doesn't need to, as that's the default. Anything that doesn't specify profiles will get Hadoop 2.2.

I believe this therefore just affects the Spark-Master-Maven-pre-YARN job in Jenkins, and certainly only affects `master`. Both of those jobs could use `-Phadoop-1`.

Looks like we don't have a Hadoop 2.2 specific build? that's fine, but then there's nothing along those lines to change.

FavioVazquez committed:
- Added missing instances in -Phadoop-1 in create-release.sh, run-tests and in the building-spark documentation

FavioVazquez commented:
Hello @srowen I was noticing some of that things. Thank you for making it easy for me to change it. I just pushed your suggested changes.

srowen commented:
ok to test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31667 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31667/consoleFull) for   PR 5786 at commit [`31bdafa`](https://github.com/apache/spark/commit/31bdafad21674fe5bc582fa678753454b04026ad).

SparkQA commented:
[Test build #31667 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31667/consoleFull) for   PR 5786 at commit [`31bdafa`](https://github.com/apache/spark/commit/31bdafad21674fe5bc582fa678753454b04026ad).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31667/
Test PASSed.

FavioVazquez commented:
@srowen @vanzin everything seems fine and test passed

shaneknapp commented:
ah cool, got it.  thanks sean!

On Sat, May 2, 2015 at 3:49 AM, Sean Owen <notifications@github.com> wrote:

> @shaneknapp <https://github.com/shaneknapp> if this goes in, then the
> essence of the config change is this: anything building for Hadoop 1.x,
> 2.0, or 2.1 now needs -Phadoop-1 to maintain the same config. Anything
> that used to specify -Phadoop-2.2 -Dhadoop.version=2.2.0 doesn't need to,
> as that's the default. Anything that doesn't specify profiles will get
> Hadoop 2.2.
>
> I believe this therefore just affects the Spark-Master-Maven-pre-YARN job
> in Jenkins, and certainly only affects master. Both of those jobs could
> use -Phadoop-1.
>
> Looks like we don't have a Hadoop 2.2 specific build? that's fine, but
> then there's nothing along those lines to change.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/5786#issuecomment-98347284>.
>

srowen commented:
Looking good. I think I will need to coordinate closely with @pwendell on this one since it would be useful to put in before the release, but also, will have some tiny implications for the release process. It finishes the process of making Hadoop 2.2 the default and that helps simplify a number of things here, so I like it.

FavioVazquez commented:
Great @srowen please let me know if I can help with something else in this patch

FavioVazquez commented:
Hello @srowen, I'm not sure about the next steps you mentioned, could you please explain me what's going to happen now with the PR. Thanks.

srowen commented:
This patch will need a rebase now, since unfortunately I merged a change that will conflict with it. It should be a simple resolution since I know the one line that will conflict. After you rebase and force-push, it's ready again.

It is still not entirely clear when/if to merge this since it will need some coordination and a release is happening. So you can sit tight.

FavioVazquez commented:
Ok I see @srowen thank you.

FavioVazquez commented:
Hello @srowen any advances in the coordination (if/when) of mergin this PR? Thanks

srowen commented:
This will take some time to review. It needs another rebase in the meantime I'm afraid. You can see that here in the UI.

pwendell commented:
Hey @FavioVazquez and @srowen. I took a look at this. A few questions:

1. Does this mean that #5027 was just wrong? I guess I don't see how things worked before this patch.
2. It's actually a pain for users when the default build changes. Why not just keep a -Phadoop-2.2 profile in the instructions? I wonder if we should just always advise users to use a Hadoop profile when building. Otherwise, we'll have to go to people and get them to change things, just like we are here. 
3. Should we just merge this into master and then just revert #5027 in branch-1.4? From what I understand the change upgrading the Hadoop version was just to make it more convenient for IDE importing. Hardly a user facing feature. Also, I think it would be good to e-mail the dev list and explain that the default build behavior is changing.

FavioVazquez commented:
I think @srowen saw this PR as a cleaned up of old dependencies and updating of spark's defaults to a currently used Hadoop version. This started as a minor fix for inconsistencies in the Hadoop defaults when using the latest CDH5 distribution, and grew to be a upgrading of the Hadoop default version, updating of docs, cleaned up yarn's POM and main POM. I still face problems when building Spark for CDH5 without this changes, and I think it would be helpful to update the versions, since Hadoop-1 is really old, and I really believe it pumps up Spark to the newest technologies.

I'm no expert in this field, but I think this PR could be interesting and useful for a lot of people that's starting with this technologies and would like to build Spark with the newest Hadoop version. I have to remark that if you use the actual building process and main POM, you'll get errors when try to connect to Cloudera's newest HDFS, yo can see that in the beginning of the PR. It's really awkward to build Spark with lots of ad hoc  and in situ dependencies just to keep old versions, Idk maybe it's just me. 

I really appreciated @srowen and @vanzin help with this, and would like to now if you think this is the right track to Spark 1.4.0 @pwendell. I'm up to making any more changes and updates if you think is necessary, and I repeat, I think this could be a good refresh to spark dependencies, I know this is really a minor change, but it could grow to be even a better update.

Thanks for your comments, I'll wait for your replies.

pwendell commented:
Hey @FavioVazquez - thanks for commenting. I'm happy to have a patch that makes the default settings more coherent. I just looked into the other pull request to understand better the origins (#5783). 

I was just a bit confused because it seems like if you were building for CDH 5.3 you would need non default settings anyways. But it seems like this was not specifically related to your issue and instead some clean-up suggested by @vainzn.

My suggestion was maybe to just make this change in master rather than putting it into the 1.4 branch, since I see basically no benefit to this other than tidiness and it introduces some natural risk of mucking around build stuff. Also, if we are going to make build changes that require developer's to build Spark differently, I think we should give ample warning. And I suggested we retain the existing profiles in our documentation in order to avoid having to keep changing developer habits every time we bump the hadoop version.

FavioVazquez commented:
I think that you guys are all right, you have suggested some great changes and I think I'll let @srowen and @vanzin, with you @pwendell decide for the future of this PR, in my humble opinion it could be good, but is all up to you guys. 

I'll be alert to the comments of this PR and please let me know if there is something I could help, making this patch better, or fixing this issues in another way. 

Thanks for teaching me great stuff.

srowen commented:
@pwendell I don't think the previous update was wrong, certainly not for development. It was insufficient for creating a Hadoop 2.2 assembly from defaults, but that's not how Hadoop 2.2 assemblies are created. In that sense, this is not required for release 1.4 to be as correct as ever.

Still the idea is that it would be better to make the default fully consistent, as if it were ready for a Hadoop 2.2 assembly.

I think the cat is out of the bag on #5027; I believe 1.3 was accidentally released with, effectively, this change? So I don't think undo that, certainly not if it's solving more problems than it causes.

(This is not at all about building for CDH.)

This doesn't remove any profiles in order to reduce impact on build scripts, yes -- otherwise `-Phadoop-2.2` would start being an error. However it must add a `hadoop-1` profile to allow selecting the Hadoop 1.x settings. This "profile" has always silently existed as the unofficial collection of defaults. Adding it does indeed require a developer change -- but only for those who need to build for Hadoop 1.x explicitly. It at least makes this explicit.

The cleanup is appealing, of course.

I would campaign modestly for introducing this into 1.4. If the above hasn't swayed your second opinion here though, then let's just do nothing for 1.4, and put this into master for 1.5. By that point I think the case will be stronger still, and there will have been time to get used to the change for the small subset of people who need to build for 1.x.

vanzin commented:
I think this is more than just a cleanup. As Favio points out, if you build with just `mvn package`, the build is broken because of inconsistent versions. The minimum command line to get a working build today is `mvn -Dhadoop.version=1.0.4 package`.

It may be that all official build scripts work around that problem inadvertently. But the current code is not correct.

So we either need to fix things so that the default profile is *actually* hadoop-2.2, or revert the previous change so that the default profile is hadoop-1. But right now the default profile is in a weird unhappy state.

srowen commented:
I don't have any problems with a plain vanilla `mvn package` ... ? what's the issue? Things that don't care about Hadoop don't care; things that do, well, sometimes do need a Hadoop dependency set to a particular version of course.

One problem here is that this version was actually already set to 2.2 in 1.3.0, at least:
https://repo1.maven.org/maven2/org/apache/spark/spark-core_2.10/1.3.1/spark-core_2.10-1.3.1.pom

If there is any issue, then I think it's best to fix-forward. I have not observed any immediate issue though?

vanzin commented:
> I don't have any problems with a plain vanilla mvn package ... ?

Can you successfully run unit tests with that? I'd expect the library version conflicts to start causing issues at that point. Also, I'm not sure the generated assembly would work on a real cluster.

srowen commented:
Yes, `mvn -DskipTests clean package; mvn test` succeeds in `master` for me now.

vanzin commented:
Just OOC, I build with no enabled profiles and tried to run spark-shell on a real cluster (standalone since YARN profile wasn't enabled). It fails pretty early with:

    15/05/12 15:08:18 INFO AppClient$ClientActor: Executor updated: app-20150512150817-0001/8 is now RUNNING
    java.lang.VerifyError: class org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos$SetOwnerRequestProto overrides final method getUnknownFields.()Lcom/google/protobuf/UnknownFieldSet;
        at java.lang.ClassLoader.defineClass1(Native Method)
        at java.lang.ClassLoader.defineClass(ClassLoader.java:800)

Which is the issue Favio raised (mismatched protobuf versions).

srowen commented:
`spark-shell` works locally for me. You're right, this may not work on Hadoop cluster X, but haven't we always generally needed to build Spark for Hadoop X to avoid this? I get it though, maybe the inconsistent Hadoop client libs don't work whereas a consistent Hadoop 1.x client lib set did, even against a mismatched cluster version.

Fair point and all that but this isn't the right way to build Spark anyway, and I'm afraid this change was effectively already released. I'm narrowly arguing against undoing the `hadoop.version=2.2.0` change. I'm also asserting that the 1.4 release artifacts will be fine.

And then saying we should fix-forward the rest of this for 1.5, if not 1.4.

vanzin commented:
> but haven't we always generally needed to build Spark for Hadoop X to avoid this?

Correct. But the default now claims to be 2.2, whereas the default build (i.e. the build where you do not enable any profiles) will not work on a hadoop-2.2 cluster.

srowen commented:
Color me surprised that the default Hadoop 1.0.4 build worked on Hadoop 2.2 (really?) though that's a fair point then. 

So, even if the `pom.xml` went back to saying Hadoop 1.0.4 for 1.4, would this still be a problem for people building against Spark 1.3? because that `pom.xml` already had this change. It seems weird to go back but not impossible.

This turns into a stronger argument for merging for 1.4 I suppose

pwendell commented:
Hey so I am okay to merge this into 1.4, but what about not having any publicly advertised "default" build and just asking people to always use profiles when building Spark in the documentation?

Otherwise every time we change the default version we are likely to make someone's life more difficult by silently changing behavior on them.

vanzin commented:
> Color me surprised that the default Hadoop 1.0.4 build worked on Hadoop 2.2 (really?) though that's a fair point then. 

I think there's some line noise somewhere. I didn't say that at all, and I wouldn't expect that to work. But that is not the issue being raised here.

> So, even if the pom.xml went back to saying Hadoop 1.0.4 for 1.4, would this still be a problem for people building against Spark 1.3?

What are you calling "a problem" here? Reverting that would revert the default build to hadoop-1. It would work on a hadoop 1 cluster, but not on a hadoop 2 cluster. Which is fine, because that's what the build would suggest.

The current problem is that the default build says it's a "hadoop-2.2" build but it does not work on a hadoop 2.2 cluster.

vanzin commented:
> but what about not having any publicly advertised "default" build

That's ok too, but it should be enforced somehow. e.g. have a profile with `activeByDefault` set to `true` than causes the build to fail with an error message.

pwendell commented:
Sure - that's good too. We can explicitly require it. But even just changing the docs to not mention default behavior would IMO be good.

srowen commented:
@vanzin what works when the build says `hadoop.version=1.0.4` that doesn't work when the build says `hadoop.version=2.2.0`? Just "running on Hadoop 1.x"? Agree but that is no longer supposed to work by default if the default Hadoop version is supposed to be 2.x. Whatever the problem is, is already a problem, since the Spark 1.3 POMs already have 2.2.0 specified.

Anyway, maybe that's just violent agreement that something has to be tweaked. If this is merged as a resolution for 1.4, OK by me for sure.

I don't like `activeByDefault` merely because it gets disabled if any profile is selected, not just a Hadoop-related profile.

I think coaching in the docs to always set these Hadoop profiles is maybe safer and more overt. Then, the net change would be: everywhere in this PR that doesn't say `-Phadoop-x.y` should add `-Phadoop-2.2`, which is actually a no-op profile, but then at least it's explicit.

Eventually when, say, Hadoop 1.x support really goes away, the `hadoop-1` profile really goes away and breaks command lines that select this profile, but, that's good.

vanzin commented:
Sean, I think you're confused. Or at least you're confusing the hell out of me.

This has nothing to do with the POMs. This has to do with the final assembly generated by the build.

If you build currently without specifying *anything* - no `-D` overrides, no profiles, no nothing - you end up with a broken build that doesn't work anywhere. That's all.

srowen commented:
We all understand that POMs define how assemblies are built. This can't have nothing to do with them.

As I say, the build works fine in, say, local mode (well, no obvious problems). Jenkins tests are happy since they don't rely on defaults. It's not true that nothing works, but there's a problem. This is narrowly about the Hadoop user's perilous expectations of defaults.

I don't expect the default assembly to work on a Hadoop 1.x cluster, but it's not supposed to now in Spark 1.4. You're suggesting reverting to restore the default to work with Hadoop 1.x, but then that trips a different version-related problem: the published POM for Spark 1.3 already references Hadoop 2.2.0. "Fixing" that may make the default assembly work for Hadoop 1.x again as it did in Spark 1.2, but then it yet again changes the transitive deps of anyone relying on Spark Core artifacts in Maven. This is why I don't think reverting to `hadoop.version=1.0.4` is a good solution, and maybe that is the only point still being batted around.

But Spark 1.4 is in a no-mans-land where the defaults don't work on 1.x (expected) and apparently don't quite work on 2.x (not expected). You'd think that at least one does. That's plainly suboptimal, and while not a show-stopper, needs fixing. I don't think anyone disputes that this PR would do the trick. Further, I like the idea of encouraging people to do the right thing, what the release has always safely done: specify Hadoop profile when it matters.

vanzin commented:
> We all understand that POMs define how assemblies are built. This can't have nothing to do with them.

You're mixing "published POMs" with the actual pom files in the build. They are not the same thing and that's what's making me incredibly confused.

>  You're suggesting reverting to restore the default to work with Hadoop 1.x

No, I'm suggesting fixing it, one way or another, which one doesn't matter.

>  but then that trips a different version-related problem: the published POM for Spark 1.3 already references Hadoop 2.2.0. 

That's completely unrelated; the build that actually does a "mvn deploy" to update the maven artifacts needs to match the profiles needed to replicate the 1.3 build. That's completely separate from what the default property values are, which is what this PR is about.

To summarize: fixing what the "default build" should be, or whether we should have a "default build" at all, is irrelevant to the published poms, and bringing them up just causes confusion.

This PR exists for a simple reason: the artifacts generated when you build with the current default properties are broken. Period.

vanzin commented:
So, I had a chat offline (well, off-github) with Sean and these are my conclusions:

- There is a real issue, addressed by this PR, that the "default build" generates an assembly that cannot talk to any version of HDFS.

- In my view, the fix proposed here is the right way forward; it standardizes on hadoop-2 as the preferred hadoop version by making it the default, and having the default build work with a hadoop 2 cluster.

- The smallest fix for the issue would be to revert back to 1.0.4 as the default version. Because we publish effective poms, that would not change the version of Hadoop for any artifacts except for spark-parent; that is not a big problem because it would only affect someone who depends on `${hadoop.version}` and has `spark-parent_2.10` as the parent project of their own project, which I'd guess is a very small set of people (if it even exists).

As for whether the "default build" should work or we should disallow it, I don't really have a strong opinion. If there's an easy fix, sure, but if it gets complicated, then it's probably not worth it.

pwendell commented:
Sounds good - however, I think the issues aren't totally decoupled, because this pull request deletes the following line from the documentation:

```
# Apache Hadoop 2.2.X
-mvn -Pyarn -Phadoop-2.2 -Dhadoop.version=2.2.0 -DskipTests clean package
```

I am suggestion look, let's just keep that line there and always suggest that people use a build profile. Otherwise changes like this will not be future proof.

Changes to default behavior are annoying for developers and I see no downside in asking people to be explicit about hadoop flags.

vanzin commented:
>  let's just keep that line there and always suggest that people use a build profile

Sure, no problems with that at all.

pwendell commented:
K - everything else from your comment makes sense, and fine to put it in 1.4. It just wasn't immediately obvious to me that it was really broken since tests and build were okay, but clearly the versions in the pom were not consistent with Hadoop 2.2.0.

FavioVazquez commented:
Perfect, I've been whatching all of your conversations. I wil make th

FavioVazquez commented:
Sorry i closed it by accident

FavioVazquez commented:
I will make the suggestes changes and push them

FavioVazquez commented:
In summary, add that line that @pwendell suggested.But I'm not sure about the default profiles, should I erase the hadoop-1 profile? there will be no default hadoop version now? Thanks

FavioVazquez commented:
And @srowen you said some days ago that you knew the places that this PR needed a Rebase, could you point them out to me please?

pwendell commented:
I think there has to be some default version by definition - and yeah let's
have the defaults at 2.2.0. But I'd just in the instructions tell people
building for hadoop 1 to use the -Phadoop1 profile and tell people building
for 2.2.0 to use the -Phadoop2.2 profile etc. I.e. let's not encourage
people to rely on the default behavior.

On Tue, May 12, 2015 at 6:31 PM, Favio André Vázquez <
notifications@github.com> wrote:

> And @srowen <https://github.com/srowen> you said some days ago that you
> knew the places that this PR needed a Rebase, could you point them out to
> me please?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/5786#issuecomment-101478579>.
>

FavioVazquez commented:
I see @pwendell, I'll push the changes tomorrow, is a little late here in Venezuela.

Greetings and thanks.

srowen commented:
@FavioVazquez Other conflicting changes have been merged to `master` since this was opened, which is why Github says "This pull request contains merge conflicts that must be resolved.". Basically you need to pull the latest `master` changes and `rebase` on it. Are you familiar with that process? I have a remote `origin` for my fork and `upstream` for the main project and so usually do ...

```
git checkout master
git pull upstream master
git checkout mybranch
git rebase master
```

You have to fix the merge conflicts and `git push origin mybranch` then.

@vanzin still disagree on the effective POM stuff; you're all correct about what should happen, but this happened in reality: 
https://repo1.maven.org/maven2/org/apache/spark/spark-core_2.10/1.3.1/spark-core_2.10-1.3.1.pom
```
<dependency>
<groupId>org.apache.hadoop</groupId>
<artifactId>hadoop-client</artifactId>
<version>2.2.0</version>
...
```
That was what I'm getting at; this shouldn't change backwards now. Whether or not that is a point doesn't matter though if we're on board with proceeding.

pwendell commented:
I think @vanzin was saying the pom for spark-parent does not have 2.2.0, it has 1.0.4. But it's moot because none of the other projects expect to get hadoop.version from it since we use effective poms.
https://repo1.maven.org/maven2/org/apache/spark/spark-parent_2.10/1.3.1/spark-parent_2.10-1.3.1.pom

But really spark-parent affects no one except for someone directly extending spark's build (which we've never expected people to do, really).

srowen commented:
Yes, agree with all that. But `spark-core` affects people. The proposed change would cause its effective POM to depend on 1.0.4 in Spark 1.4, assuming it's published as intended with no build flags, whereas Spark 1.3 depends on 2.2.0 (which I understand was inadvertent to begin with).

Is the disconnect that people think that's fine? I'd be surprised, since actual, real Hadoop dependency for Spark Core would have flip flopped from 2.2.0 to 1.0.4 and back to 2.2.0 over three releases.

pwendell commented:
I thought the proposal here was to continue publishing artifacts with 2.2.0 (?) If you look at the patch, it moves the default build to 2.2.0 and then it publishes with the default build. I think that's the way to go... but maybe I'm misunderstanding?

srowen commented:
Yes, I'm arguing that the alternative, changing `hadoop.version` back to 1.0.4, is not a solution. I disagree with "The smallest fix for the issue would be to revert back to 1.0.4 as the default version" -- a fix perhaps for one thing here, but that causes a different problem. I am in favor of this PR.

pwendell commented:
Oh yeah sorry - I was wrong about that. Never said so explicitly, but basically I support this patch in it's current form (brought up to date) with the only change being to not ask users to rely on default behavior in our docs (i.e. don't delete the docs referring to hadoop-2.2).

srowen commented on dev/create-release/create-release.sh:
```diff
@@ -118,14 +118,14 @@ if [[ ! "$@" =~ --skip-publish ]]; then
 
-  build/mvn -DskipTests -Dhadoop.version=2.2.0 -Dyarn.version=2.2.0 \
-    -Pyarn -Phive -Phive-thriftserver -Phadoop-2.2 -Pspark-ganglia-lgpl -Pkinesis-asl \
+  build/mvn -DskipTests -Pyarn -Phive \
```
Favio I think these two now need `-Phadoop-2.2` again to be fully consistent.

srowen commented on dev/run-tests:
```diff
@@ -40,11 +40,11 @@ function handle_error () {
+      export SBT_MAVEN_PROFILES_ARGS="-Phadoop-1 -Dhadoop.version=2.0.0-mr1-cdh4.1.1"
     elif [ "$AMPLAB_JENKINS_BUILD_PROFILE" = "hadoop2.2" ]; then
-      export SBT_MAVEN_PROFILES_ARGS="-Pyarn -Phadoop-2.2 -Dhadoop.version=2.2.0"
+      export SBT_MAVEN_PROFILES_ARGS="-Pyarn"
```
`-Phadoop-2.2` here

srowen commented on dev/scalastyle:
```diff
@@ -20,8 +20,8 @@
 # Check style with YARN built too
-echo -e "q\n" | build/sbt -Pyarn -Phadoop-2.2 -Dhadoop.version=2.2.0 scalastyle >> scalastyle.txt
-echo -e "q\n" | build/sbt -Pyarn -Phadoop-2.2 -Dhadoop.version=2.2.0 test:scalastyle >> scalastyle.txt
+echo -e "q\n" | build/sbt -Pyarn scalastyle >> scalastyle.txt
```
`-Phadoop-2.2` here

srowen commented on docs/building-spark.md:
```diff
@@ -67,8 +67,8 @@ Because HDFS is not protocol-compatible across versions, if you want to read fro
-    <tr><td>1.x to 2.1.x</td><td>(none)</td></tr>
-    <tr><td>2.2.x</td><td>hadoop-2.2</td></tr>
+    <tr><td>1.x to 2.1.x</td><td>hadoop-1</td></tr>
+    <tr><td>2.2.x</td><td>(none)</td></tr>
```
And back here...

srowen commented on docs/building-spark.md:
```diff
@@ -92,8 +92,6 @@ You can enable the "yarn" profile and optionally set the "yarn.version" property
 Examples:
 
 {% highlight bash %}
-# Apache Hadoop 2.2.X
```
This can come back

FavioVazquez commented:
Great, I'm familiar with the process @srowen. Thank you guys for all the suggestions, I'm making the changes and be pushing the changes soon

FavioVazquez committed:
Merge branch 'master' of https://github.com/apache/spark into update-hadoop-dependencies

Conflicts:
	docs/building-spark.md
	make-distribution.sh
	pom.xml

FavioVazquez committed:
- Added instances of -Phadoop-2.2 in create-release.sh, run-tests, scalastyle and building-spark.md
- Reconstructed docs to not ask users to rely on default behavior

FavioVazquez commented:
I hope that this is the correct way of making all the changes you suggested. Please check this and thank you @srowen @vanzin and @pwendell. Let me know if there is something else that could be done, or if this finishes the patch.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32605 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32605/consoleFull) for   PR 5786 at commit [`379f50d`](https://github.com/apache/spark/commit/379f50d63629318d1d0689a155a201a220aa54fe).

SparkQA commented:
[Test build #32605 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32605/consoleFull) for   PR 5786 at commit [`379f50d`](https://github.com/apache/spark/commit/379f50d63629318d1d0689a155a201a220aa54fe).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32605/
Test PASSed.

srowen commented on dev/create-release/create-release.sh:
```diff
@@ -118,14 +118,14 @@ if [[ ! "$@" =~ --skip-publish ]]; then
-  build/mvn -DskipTests -Dhadoop.version=2.2.0 -Dyarn.version=2.2.0 \
-    -Dscala-2.11 -Pyarn -Phive -Phadoop-2.2 -Pspark-ganglia-lgpl -Pkinesis-asl \
+  build/mvn -DskipTests -Pyarn -Phive \
+    -Dscala-2.11 -Pspark-ganglia-lgpl -Pkinesis-asl \
```
This still needs `-Phadoop-2.2` but maybe I can add that on merge

FavioVazquez commented:
Are you sure Sean? I could make the change and push it, but if is easier to make the change in the merge you tell me.

srowen commented:
Go ahead if you have a moment; only if it's not much work.
Thanks for your perseverance. This ends up being a great change IMHO. It will go in shortly.

FavioVazquez commented:
I'm happy to help, give me a sec and I'll push the changes

FavioVazquez committed:
- Added missing instance of -Phadoop-2.2 in create-release.sh

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32700 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32700/consoleFull) for   PR 5786 at commit [`11670e5`](https://github.com/apache/spark/commit/11670e5baf49489c2d0e394a32865deff8e3a791).

SparkQA commented:
[Test build #32700 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32700/consoleFull) for   PR 5786 at commit [`11670e5`](https://github.com/apache/spark/commit/11670e5baf49489c2d0e394a32865deff8e3a791).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32700/
Test FAILed.

FavioVazquez commented:
This has happened before @srowen, I think this is again an unrelated fail. Could you ask jenkins to retest this please?

srowen commented:
I'm sure it is as you only made a doc change but while we are waiting: Jenkins retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32708 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32708/consoleFull) for   PR 5786 at commit [`11670e5`](https://github.com/apache/spark/commit/11670e5baf49489c2d0e394a32865deff8e3a791).

SparkQA commented:
[Test build #32708 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32708/consoleFull) for   PR 5786 at commit [`11670e5`](https://github.com/apache/spark/commit/11670e5baf49489c2d0e394a32865deff8e3a791).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32708/
Test PASSed.

FavioVazquez commented:
All test passed @srowen. It was, as expected, an unrelated error. Is everything set now to merge this PR?

shaneknapp commented:
@srowen -- so i'm looking at the configs and we don't actually specify anything in the builds WRT to mvn or sbt build options[1].  these are all matrix build configs and the hadoop options are all handled in the scripts in dev/... (for the most part).

[1] - not *technically* true, but these guys don't seem like they'll break as the hadoop version > 2.2.0:
  - https://amplab.cs.berkeley.edu/jenkins/job/Spark-1.4-Maven-with-YARN/configure
  - https://amplab.cs.berkeley.edu/jenkins/job/Spark-1.3-Maven-with-YARN/configure
  - https://amplab.cs.berkeley.edu/jenkins/job/Spark-1.2-Maven-with-YARN/configure
